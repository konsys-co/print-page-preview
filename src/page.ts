type PageSize = 'A4' | 'A4LSC' | 'A5' | 'A5LSC' | 'CUSTOM'

type PageConfig = {
  pageSize: PageSize | string
  width?: number
  height?: number
}

const getWidthHeight = (config: PageConfig): { width: number, height: number } => {
  const { pageSize, width, height } = config
  switch (pageSize) {
    case 'A4' :
      return { width: 210, height: 297 }
    case 'A4LSC' :
      return { width: 297, height: 210 }
    case 'A5' :
      return { width: 148, height: 210 }
    case 'A5LSC' :
      return { width: 210, height: 148 }
    case 'CUSTOM' :
      if (!width || !height) throw new Error('Custom size must have width and height')
      return { width, height }
    default :
      return { width: 210, height: 297 }
  }
}

class PageTemplate {
  pages: { content: string, customWidth?: number, customHeight?: number }[]
  title: string
  heads: string
  styles: string

  constructor(
    pages: { content: string, customConfig?: PageConfig }[],
    options: { title?: string, heads?: string, styles?: string } = {}
  ) {
    this.pages = pages.map(page => {
      const { content, customConfig } = page
      if (customConfig) {
        const { width, height } = getWidthHeight(customConfig)
        return { content, customWidth: width, customHeight: height }
      }
      return { content }
    })
    this.title = options.title || 'No Title'
    this.heads = options.heads || ''
    this.styles = options.styles || ''
  }

  public generatePrintablePage(defaultConfig: PageConfig = { pageSize: 'A4' }) {
    const { width, height } = getWidthHeight(defaultConfig)
    return `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>${this.title}</title>
    ${this.heads}
    <style>
      :root {
        --bleeding: 1cm;
      }
      * {
        box-sizing: border-box;
      }
      body {
        margin: 0 auto;
        padding: 0;
        background: rgb(204, 204, 204);
        display: flex;
        flex-direction: column;
      }
      b {
        font-weight: 600;
      }
      .page {
        display: inline-block;
        position: relative;
        margin: 2em auto;
        padding: var(--bleeding);
        box-shadow: 0 0 0 rgba(0, 0, 0, 0.5);
        background: white;
        font-family: Arial, Helvetica, sans-serif;
      }
      ${
        this.pages.map((page, i) => `
        @page page-${i} {
          size: ${page.customWidth || width}mm ${page.customHeight || height}mm;
          margin: 0;
        }
        .page-${i} {
          width: ${page.customWidth || width}mm;
          height: ${page.customHeight || height}mm;
        }
        `
        ).join('')
      }
      @media screen {
        .page::after {
          position: absolute;
          content: "";
          top: 0;
          left: 0;
          width: calc(100% - var(--bleeding) * 2);
          height: calc(100% - var(--bleeding) * 2);
          margin: var(--bleeding);
          outline: thin dashed rgba(0, 0, 0, 0.2);
          pointer-events: none;
          z-index: 9999;
        }
      }
      @media print {
        .page {
          margin: 0;
          overflow: hidden;
        }
        ${
          this.pages.map((_, i) => `
          .page-${i} {
            page: page-${i};
          }
          `
          ).join('')
        }
      }
      ${this.styles}
    </style>
  </head>
  <body>
    ${this.pages.map((page, i) => `<div class="page page-${i}">${page.content}</div>`).join('')}
  </body>
</html>
    `
  }
}

export { PageTemplate }