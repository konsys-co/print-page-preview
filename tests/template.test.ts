import { PageTemplate } from '../src'

test('Default A4', () => {
  const pages = [
    { content: 'Page 1' },
    { content: 'Page 2' },
    { content: 'Page 3' },
  ]
  const pageTemplate = new PageTemplate(pages)
  const printablePage = pageTemplate.generatePrintablePage()
  expect(printablePage).toMatchSnapshot()
})

test('Different cover page', () => {
  const pages = [
    { content: 'Cover Page', customConfig: { pageSize: 'A5LSC' } },
    { content: 'Page 1' },
    { content: 'Page 2' },
  ]
  const options = {
    title: 'Custom Cover Page',
    heads: '<style>body { font-family: sans-serif; }</style>',
    styles: 'body { background-color: #f0f0f0; }',
  }
  const pageTemplate = new PageTemplate(pages, options)
  const printablePage = pageTemplate.generatePrintablePage({ pageSize: 'A5' })
  expect(printablePage).toMatchSnapshot()
})

test('Custom size page', () => {
  const pages = [
    { content: 'Page 1' },
    { content: 'Page 2' },
    { content: 'Page 3' },
  ]
  const options = {
    title: 'Custom Size Page',
    heads: '<style>body { font-family: sans-serif; }</style>',
    styles: 'body { background-color: #f0f0f0; }',
  }
  const pageTemplate = new PageTemplate(pages, options)
  const printablePage = pageTemplate.generatePrintablePage({ pageSize: 'CUSTOM', width: 200, height: 300 })
  expect(printablePage).toMatchSnapshot()
})

test('Error on custom size page without width and height', () => {
  const pages = [
    { content: 'Custom Size Page', customConfig: { pageSize: 'CUSTOM' } },
    { content: 'Page 1' },
    { content: 'Page 2' },
  ]
  const options = {
    title: 'Custom Size Page',
    heads: '<style>body { font-family: sans-serif; }</style>',
    styles: 'body { background-color: #f0f0f0; }',
  }
  expect(() => new PageTemplate(pages, options)).toThrowError('Custom size must have width and height')
})