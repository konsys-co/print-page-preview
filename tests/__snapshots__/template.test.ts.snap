// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom size page 1`] = `
"
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Custom Size Page</title>
    <style>body { font-family: sans-serif; }</style>
    <style>
      :root {
        --bleeding: 1cm;
      }
      * {
        box-sizing: border-box;
      }
      body {
        margin: 0 auto;
        padding: 0;
        background: rgb(204, 204, 204);
        display: flex;
        flex-direction: column;
      }
      b {
        font-weight: 600;
      }
      .page {
        display: inline-block;
        position: relative;
        margin: 2em auto;
        padding: var(--bleeding);
        box-shadow: 0 0 0 rgba(0, 0, 0, 0.5);
        background: white;
        font-family: Arial, Helvetica, sans-serif;
      }
      
        @page page-0 {
          size: 200mm 300mm;
          margin: 0;
        }
        .page-0 {
          width: 200mm;
          height: 300mm;
        }
        
        @page page-1 {
          size: 200mm 300mm;
          margin: 0;
        }
        .page-1 {
          width: 200mm;
          height: 300mm;
        }
        
        @page page-2 {
          size: 200mm 300mm;
          margin: 0;
        }
        .page-2 {
          width: 200mm;
          height: 300mm;
        }
        
      @media screen {
        .page::after {
          position: absolute;
          content: "";
          top: 0;
          left: 0;
          width: calc(100% - var(--bleeding) * 2);
          height: calc(100% - var(--bleeding) * 2);
          margin: var(--bleeding);
          outline: thin dashed rgba(0, 0, 0, 0.2);
          pointer-events: none;
          z-index: 9999;
        }
      }
      @media print {
        .page {
          margin: 0;
          overflow: hidden;
        }
        
          .page-0 {
            page: page-0;
          }
          
          .page-1 {
            page: page-1;
          }
          
          .page-2 {
            page: page-2;
          }
          
      }
      body { background-color: #f0f0f0; }
    </style>
  </head>
  <body>
    <div class="page page-0">Page 1</div><div class="page page-1">Page 2</div><div class="page page-2">Page 3</div>
  </body>
</html>
    "
`;

exports[`Default A4 1`] = `
"
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>No Title</title>
    
    <style>
      :root {
        --bleeding: 1cm;
      }
      * {
        box-sizing: border-box;
      }
      body {
        margin: 0 auto;
        padding: 0;
        background: rgb(204, 204, 204);
        display: flex;
        flex-direction: column;
      }
      b {
        font-weight: 600;
      }
      .page {
        display: inline-block;
        position: relative;
        margin: 2em auto;
        padding: var(--bleeding);
        box-shadow: 0 0 0 rgba(0, 0, 0, 0.5);
        background: white;
        font-family: Arial, Helvetica, sans-serif;
      }
      
        @page page-0 {
          size: 210mm 297mm;
          margin: 0;
        }
        .page-0 {
          width: 210mm;
          height: 297mm;
        }
        
        @page page-1 {
          size: 210mm 297mm;
          margin: 0;
        }
        .page-1 {
          width: 210mm;
          height: 297mm;
        }
        
        @page page-2 {
          size: 210mm 297mm;
          margin: 0;
        }
        .page-2 {
          width: 210mm;
          height: 297mm;
        }
        
      @media screen {
        .page::after {
          position: absolute;
          content: "";
          top: 0;
          left: 0;
          width: calc(100% - var(--bleeding) * 2);
          height: calc(100% - var(--bleeding) * 2);
          margin: var(--bleeding);
          outline: thin dashed rgba(0, 0, 0, 0.2);
          pointer-events: none;
          z-index: 9999;
        }
      }
      @media print {
        .page {
          margin: 0;
          overflow: hidden;
        }
        
          .page-0 {
            page: page-0;
          }
          
          .page-1 {
            page: page-1;
          }
          
          .page-2 {
            page: page-2;
          }
          
      }
      
    </style>
  </head>
  <body>
    <div class="page page-0">Page 1</div><div class="page page-1">Page 2</div><div class="page page-2">Page 3</div>
  </body>
</html>
    "
`;

exports[`Different cover page 1`] = `
"
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Custom Cover Page</title>
    <style>body { font-family: sans-serif; }</style>
    <style>
      :root {
        --bleeding: 1cm;
      }
      * {
        box-sizing: border-box;
      }
      body {
        margin: 0 auto;
        padding: 0;
        background: rgb(204, 204, 204);
        display: flex;
        flex-direction: column;
      }
      b {
        font-weight: 600;
      }
      .page {
        display: inline-block;
        position: relative;
        margin: 2em auto;
        padding: var(--bleeding);
        box-shadow: 0 0 0 rgba(0, 0, 0, 0.5);
        background: white;
        font-family: Arial, Helvetica, sans-serif;
      }
      
        @page page-0 {
          size: 210mm 148mm;
          margin: 0;
        }
        .page-0 {
          width: 210mm;
          height: 148mm;
        }
        
        @page page-1 {
          size: 148mm 210mm;
          margin: 0;
        }
        .page-1 {
          width: 148mm;
          height: 210mm;
        }
        
        @page page-2 {
          size: 148mm 210mm;
          margin: 0;
        }
        .page-2 {
          width: 148mm;
          height: 210mm;
        }
        
      @media screen {
        .page::after {
          position: absolute;
          content: "";
          top: 0;
          left: 0;
          width: calc(100% - var(--bleeding) * 2);
          height: calc(100% - var(--bleeding) * 2);
          margin: var(--bleeding);
          outline: thin dashed rgba(0, 0, 0, 0.2);
          pointer-events: none;
          z-index: 9999;
        }
      }
      @media print {
        .page {
          margin: 0;
          overflow: hidden;
        }
        
          .page-0 {
            page: page-0;
          }
          
          .page-1 {
            page: page-1;
          }
          
          .page-2 {
            page: page-2;
          }
          
      }
      body { background-color: #f0f0f0; }
    </style>
  </head>
  <body>
    <div class="page page-0">Cover Page</div><div class="page page-1">Page 1</div><div class="page page-2">Page 2</div>
  </body>
</html>
    "
`;
